File: Sell_course_server/src/main/java/com/study/sell_course/controller/CourseController.java
package com.study.sell_course.controller;

import com.study.sell_course.dto.CourseDTO;
import com.study.sell_course.service.CourseService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/courses")
@Tag(name = "Course APIs")
public class CourseController {

    @Autowired
    private CourseService courseService;

    @Operation(summary = "Get all courses", description = "Retrieve a list of all courses")
    @ApiResponse(responseCode = "200", description = "Successful retrieval of course list")
    @GetMapping
    public ResponseEntity<List<CourseDTO>> getAllCourses() {
        return ResponseEntity.ok(courseService.getAllCourses());
    }

    @Operation(summary = "Get a course by ID", description = "Retrieve course details by course ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successful retrieval of course details"),
            @ApiResponse(responseCode = "404", description = "Course not found", content = @Content)
    })
    @GetMapping("/{courseId}")
    public ResponseEntity<CourseDTO> getCourseById(@PathVariable String courseId) {
        return ResponseEntity.ok(courseService.getCourseById(courseId));
    }

    @Operation(summary = "Create a new course", description = "Add a new course to the system")
    @ApiResponse(responseCode = "200", description = "Course successfully created")
    @PostMapping
    public ResponseEntity<CourseDTO> createCourse(@RequestBody CourseDTO courseDTO) {
        return ResponseEntity.ok(courseService.createCourse(courseDTO));
    }

    @Operation(summary = "Update a course", description = "Update details of an existing course by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Course successfully updated"),
            @ApiResponse(responseCode = "404", description = "Course not found", content = @Content)
    })
    @PutMapping("/{courseId}")
    public ResponseEntity<CourseDTO> updateCourse(@PathVariable String courseId, @RequestBody CourseDTO courseDTO) {
        return ResponseEntity.ok(courseService.updateCourse(courseId, courseDTO));
    }

    @Operation(summary = "Delete a course", description = "Remove a course from the system by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "204", description = "Course successfully deleted"),
            @ApiResponse(responseCode = "404", description = "Course not found", content = @Content)
    })
    @DeleteMapping("/{courseId}")
    public ResponseEntity<Void> deleteCourse(@PathVariable String courseId) {
        courseService.deleteCourse(courseId);
        return ResponseEntity.noContent().build();
    }
}

File: sell_course_nest/src/modules/course/course.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Put,
  Patch,
  UploadedFiles,
  UseInterceptors,
} from '@nestjs/common';
import { CourseRequestDTO } from './dto/courseRequestData.dto';
import { CourseResponseDTO } from './dto/courseResponseData.dto';
import { CourseDetailResponse } from './dto/courseDetailResponse.dto';
import {
  UpdateCourseStatusDto,
  ReviewCourseStatusDto,
} from './dto/update-course-status.dto';
import { CourseStatus } from './enums/course-status.enum';
import { CourseService } from './course.service';
import {
  ApiBearerAuth,
  ApiConsumes,
  ApiOperation,
  ApiResponse,
} from '@nestjs/swagger';
import { FileFieldsInterceptor } from '@nestjs/platform-express';
@Controller('api')
export class CourseController {
  constructor(private readonly courseService: CourseService) {}

  @Get('courses/getAll')
  @ApiOperation({ summary: 'Get all courses' })
  @ApiResponse({
    status: 200,
    description: 'Successfully retrieved all courses.',
    type: [CourseResponseDTO],
  })
  @ApiResponse({
    status: 404,
    description: 'No courses found.',
  })
  async getAllCourses(): Promise<CourseResponseDTO[]> {
    return await this.courseService.getAllCourses();
  }

  @Get('instructor/courses/view_course')
  @ApiBearerAuth('Authorization')
  @ApiOperation({ summary: 'Get all courses' })
  @ApiResponse({
    status: 200,
    description: 'Successfully retrieved all courses.',
    type: [CourseResponseDTO],
  })
  @ApiResponse({
    status: 404,
    description: 'No courses found.',
  })
  async getAllCoursesAdmin(): Promise<CourseResponseDTO[]> {
    return await this.courseService.getAllCourses();
  }

  @ApiBearerAuth('Authorization')
  @Get('instructor/courses/view_course/:id')
  @ApiBearerAuth('Authorization')
  @ApiOperation({ summary: 'Get course by ID' })
  @ApiResponse({
    status: 200,
    description: 'Successfully retrieved the course.',
    type: CourseResponseDTO,
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found with the given ID.',
  })
  async getCourseByIdAmin(
    @Param('id') courseId: string,
  ): Promise<CourseResponseDTO> {
    return await this.courseService.getCourseById(courseId);
  }

  @Get('courses/getByCourse/:id')
  @ApiOperation({ summary: 'Get course by ID' })
  @ApiResponse({
    status: 200,
    description: 'Successfully retrieved the course.',
    type: CourseResponseDTO,
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found with the given ID.',
  })
  async getCourseById(
    @Param('id') courseId: string,
  ): Promise<CourseResponseDTO> {
    return await this.courseService.getCourseById(courseId);
  }

  @ApiBearerAuth('Authorization')
  @Post('instructor/courses/create_course')
  @UseInterceptors(
    FileFieldsInterceptor([
      { name: 'videoIntro', maxCount: 1 },
      { name: 'thumbnail', maxCount: 1 },
    ]),
  )
  @ApiConsumes('multipart/form-data')
  @ApiOperation({ summary: 'Create a new course' })
  @ApiResponse({
    status: 201,
    description: 'The course has been successfully created.',
    type: CourseResponseDTO,
  })
  @ApiResponse({ status: 400, description: 'Invalid course data provided.' })
  async createCourse(
    @Body() course: CourseRequestDTO,
    @UploadedFiles()
    files?: {
      videoIntro?: Express.Multer.File[];
      thumbnail?: Express.Multer.File[];
    },
  ): Promise<CourseResponseDTO> {
    return await this.courseService.createCourse(course, files ?? {});
  }

  @ApiBearerAuth('Authorization')
  @Put('/instructor/courses/update_course/:id')
  @UseInterceptors(
    FileFieldsInterceptor([
      { name: 'videoIntro', maxCount: 1 },
      { name: 'thumbnail', maxCount: 1 },
    ]),
  )
  @ApiConsumes('multipart/form-data')
  @ApiOperation({ summary: 'Update an existing course' })
  @ApiResponse({
    status: 200,
    description: 'The course has been successfully updated.',
    type: CourseResponseDTO,
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found with the given ID.',
  })
  @ApiResponse({ status: 400, description: 'Invalid course data provided.' })
  async updateCourse(
    @Param('id') courseId: string,
    @Body() updateData: CourseRequestDTO,
    @UploadedFiles()
    files?: {
      videoIntro?: Express.Multer.File[];
      thumbnail?: Express.Multer.File[];
    },
  ): Promise<CourseResponseDTO> {
    return await this.courseService.updateCourse(
      courseId,
      updateData,
      files ?? {},
    );
  }

  @ApiBearerAuth('Authorization')
  @Delete('instructor/courses/delete_course/:id')
  @ApiOperation({ summary: 'Delete a course by ID' })
  @ApiResponse({
    status: 200,
    description: 'The course has been successfully deleted.',
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found with the given ID.',
  })
  async deleteCourse(@Param('id') courseId: string): Promise<void> {
    await this.courseService.deleteCourse(courseId);
  }

  @Get('courses/details/:id')
  @ApiOperation({ summary: 'Get course details with lessons and contents' })
  @ApiResponse({
    status: 200,
    description:
      'Successfully retrieved course details with lessons and contents.',
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found with the given ID.',
  })
  async getCourseDetails(
    @Param('id') courseId: string,
  ): Promise<CourseDetailResponse> {
    return this.courseService.getCourseWithLessonsAndContents(courseId);
  }

  @Get('getByCategory/:category_id')
  async getByCategory(@Param('category_id') category_id: string) {
    return this.courseService.getCoursesByCategory(category_id);
  }

  @Patch('courses/:courseId/status')
  @ApiBearerAuth('Authorization')
  @ApiOperation({
    summary: 'Update course status (instructor only - DRAFT/PENDING_REVIEW)',
  })
  @ApiResponse({
    status: 200,
    description: 'Course status updated successfully.',
  })
  @ApiResponse({
    status: 400,
    description: 'Invalid status transition.',
  })
  @ApiResponse({
    status: 403,
    description: 'You can only update status of your own courses.',
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found.',
  })
  async updateCourseStatus(
    @Param('courseId') courseId: string,
    @Body() updateStatusDto: UpdateCourseStatusDto,
    // TODO: Get instructorId from JWT token in authentication guard
    // For now, you'll need to add authentication and extract from token
  ): Promise<{ message: string }> {
    // This should be extracted from JWT token in real implementation
    const instructorId = 'temp-instructor-id'; // Replace with actual JWT extraction
    return this.courseService.updateCourseStatus(
      courseId,
      updateStatusDto,
      instructorId,
    );
  }

  @Patch('admin/courses/:courseId/review')
  @ApiBearerAuth('Authorization')
  @ApiOperation({
    summary: 'Review course status (admin only - PUBLISHED/REJECTED)',
  })
  @ApiResponse({
    status: 200,
    description: 'Course reviewed successfully.',
  })
  @ApiResponse({
    status: 400,
    description: 'Only courses in PENDING_REVIEW status can be reviewed.',
  })
  @ApiResponse({
    status: 404,
    description: 'Course not found.',
  })
  async reviewCourseStatus(
    @Param('courseId') courseId: string,
    @Body() reviewStatusDto: ReviewCourseStatusDto,
  ): Promise<{ message: string }> {
    return this.courseService.reviewCourseStatus(courseId, reviewStatusDto);
  }

  @Get('admin/courses/status/:status')
  @ApiBearerAuth('Authorization')
  @ApiOperation({ summary: 'Get courses by status (admin only)' })
  @ApiResponse({
    status: 200,
    description: 'Successfully retrieved courses by status.',
    type: [CourseResponseDTO],
  })
  async getCoursesByStatus(
    @Param('status') status: CourseStatus,
  ): Promise<CourseResponseDTO[]> {
    return this.courseService.getCoursesByStatus(status);
  }
}
